### ISSUE TITLE
Reproduced Results Inconsistent with Paper

### ISSUE THREAD

**Issue by xingyifei2016** [h-zhao1997/cobra#11]
As the authors do not release the evaluation code, I have personally written code for evaluation via adaptation of LLaVa in the GQA task. However, my reproduced result from the official checkpoint (cobra+3b) does not align with the reported results, only at around 45% accuracy as opposed to the 58% reported.

Binary: 69.34%
Open: 25.53%
Accuracy: 45.64%
Validity: 0.00%
Plausibility: 0.00%
Distribution: 2.62 (lower is better)

Accuracy / structural type:
  choose: 62.09% (1129 questions)
  compare: 56.37% (589 questions)
  logical: 69.77% (1803 questions)
  query: 25.53% (6805 questions)
  verify: 76.02% (2252 questions)

Accuracy / semantic type:
  attr: 55.88% (5186 questions)
  cat: 21.06% (1149 questions)
  global: 38.85% (157 questions)
  obj: 83.03% (778 questions)
  rel: 35.66% (5308 questions)

Accuracy / steps number:
  1: 50.21% (237 questions)
  2: 37.20% (6395 questions)
  3: 50.82% (4266 questions)
  4: 58.01% (793 questions)
  5: 67.76% (822 questions)
  6: 85.37% (41 questions)
  7: 95.00% (20 questions)
  8: 66.67% (3 questions)
  9: 100.00% (1 questions)

Accuracy / words number:
  3: 27.15% (151 questions)
  4: 33.97% (630 questions)
  5: 32.64% (1290 questions)
  6: 39.87% (2074 questions)
  7: 47.02% (1642 questions)
  8: 46.50% (1185 questions)
  9: 51.44% (1281 questions)
  10: 52.04% (1249 questions)
  11: 49.20% (994 questions)
  12: 51.25% (638 questions)
  13: 48.05% (462 questions)
  14: 56.81% (345 questions)
  15: 61.18% (237 questions)
  16: 52.99% (117 questions)
  17: 54.26% (94 questions)
  18: 56.58% (76 questions)
  19: 74.42% (43 questions)
  20: 50.00% (32 questions)
  21: 68.42% (19 questions)
  22: 50.00% (12 questions)
  23: 25.00% (4 questions)
  24: 50.00% (2 questions)
  25: 100.00% (1 questions)

**Comment by xingyifei2016** [comment]
These are results from the actual checkpoints released by the authors. If I use the script to pretrain from scratch, the performance on gqa further decreases to this:

Binary: 35.98%
Open: 2.29%
Accuracy: 17.75%
Validity: 0.00%
Plausibility: 0.00%
Distribution: 3.96 (lower is better)

Accuracy / structural type:
  choose: 24.89% (1129 questions)
  compare: 27.50% (589 questions)
  logical: 44.32% (1803 questions)
  query: 2.29% (6805 questions)
  verify: 37.08% (2252 questions)

Accuracy / semantic type:
  attr: 20.30% (5186 questions)
  cat: 5.92% (1149 questions)
  global: 18.47% (157 questions)
  obj: 65.68% (778 questions)
  rel: 10.78% (5308 questions)

Accuracy / steps number:
  1: 27.00% (237 questions)
  2: 10.41% (6395 questions)
  3: 21.21% (4266 questions)
  4: 30.01% (793 questions)
  5: 39.66% (822 questions)
  6: 51.22% (41 questions)
  7: 55.00% (20 questions)
  8: 66.67% (3 questions)
  9: 0.00% (1 questions)

Accuracy / words number:
  3: 5.30% (151 questions)
  4: 8.89% (630 questions)
  5: 8.76% (1290 questions)
  6: 14.22% (2074 questions)
  7: 21.68% (1642 questions)
  8: 21.77% (1185 questions)
  9: 20.92% (1281 questions)
  10: 19.78% (1249 questions)
  11: 21.03% (994 questions)
  12: 20.38% (638 questions)
  13: 17.75% (462 questions)
  14: 24.64% (345 questions)
  15: 27.43% (237 questions)
  16: 17.95% (117 questions)
  17: 14.89% (94 questions)
  18: 11.84% (76 questions)
  19: 11.63% (43 questions)
  20: 9.38% (32 questions)
  21: 31.58% (19 questions)
  22: 0.00% (12 questions)
  23: 25.00% (4 questions)
  24: 50.00% (2 questions)
  25: 100.00% (1 questions)

**Comment by h-zhao1997** [comment]
Since I do not have access to your code, I am unable to determine the reason for the mismatch in the reproduced results. However, we have released the evaluation code that we used, which can be found at https://github.com/h-zhao1997/vlm-evaluation.

**Comment by WenjunHuang94** [comment]
> Since I do not have access to your code, I am unable to determine the reason for the mismatch in the reproduced results. However, we have released the evaluation code that we used, which can be found at https://github.com/h-zhao1997/vlm-evaluation.

How can I resolve an error message when using the pip install - e. command to install vlm-evaluation? 
![image](https://github.com/h-zhao1997/cobra/assets/20516638/ba726a35-1422-47ec-9a8a-fce52f8b7948)

The library I installed is currently as follows:
![image](https://github.com/h-zhao1997/cobra/assets/20516638/58803937-fd50-4c14-b22d-ec609335b013)

**Comment by xingyifei2016** [comment]
Thank you, Han. Given your code, I have reproduced the results once again. The released model indeed matches with the paper. 
![WechatIMG364](https://github.com/h-zhao1997/cobra/assets/25236798/d04b5474-c0c4-4346-8c43-157f0e73fd46)
However, if I use your script to run a model (for 2 epochs), the reproduced results are clearly worse.
![WechatIMG457](https://github.com/h-zhao1997/cobra/assets/25236798/8fef023c-b540-4418-b674-2c013dd7fa1d)
Here, I have modified the original script at only one place: I changed finetune_per_device_batch_size from 8 to 4, due to hardware limitations (running using 40G A100s). How is this happening? Also, I noticed that despite setting a seed, the script produces models that differ in final parameters after 2 epochs, as shown here with multiple runs.
![WechatIMG383](https://github.com/h-zhao1997/cobra/assets/25236798/da0ca49f-d0de-4e0b-acd0-f17193fa6204)
Please advise.

**Comment by h-zhao1997** [comment]
@xingyifei2016 We have not tested the training process on the 4-card setup, so I am currently unable to determine the cause. As for the issue of randomness, it involves the sampler, the model, and the libraries that are related to randomness (random, torch, numpy), so it's a bit strange that this phenomenon occurs, and I am also unable to determine the cause on your end.

**Comment by xingyifei2016** [comment]
No, it is still in eight card training setting, but with reduced storage, 40G compared to 80G A100s you have.

**Comment by Liuxueyi** [comment]
Hi, did you compare the inference time of these models, and how? Looking forward to your reply.
